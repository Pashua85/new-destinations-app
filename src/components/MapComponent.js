import React from 'react';
import { compose, withProps } from 'recompose';
import { 
  withGoogleMap, 
  GoogleMap, 
  Marker, 
  Polyline, 
  InfoWindow
} from 'react-google-maps';

class Map extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showInfoBox: false,
      selectPoint: null    
    };
  };

  componentDidUpdate() {
    if(this.props.points.length > 1) {
      const bounds = new window.google.maps.LatLngBounds();
      this.props.points.map((point, i) => {
        bounds.extend(new window.google.maps.LatLng(
          point.latLng.lat,
          point.latLng.lng
        ));
      });
      this.map.fitBounds(bounds);
    };
  }

  resetState = () => {
    this.setState({ 
      showInfoBox: false,
      selectPoint: null
    });    
  }

  handleClickedMarker = (point) => () => {
    this.setState(() => ({
      selectPoint: point,
      showInfoBox: true
    }));
  };

  onStartChangePosition = ()  => {
    this.setState(() => ({
      showInfoBox: false
    }))
  };

  handleCenterChanged = () => {
    const currentCenter = {
      lat: this.map.getCenter().lat(),
      lng: this.map.getCenter().lng()
    };
    this.props.onCenterChanged(currentCenter);
  };

  handleCloseInfoBox = () => {
    this.setState({ 
      selectPoint: null,
      showInfoBox: false
    });
  };

  render() {   
    var center;
    this.props.points.length ?
    center = this.props.points[0].latLng :
    center = {
      lat: 55.751244,
      lng: 37.618423
    };


    return (
      <GoogleMap
        ref={ref => { this.map = ref;}}
        defaultZoom={14}  
        defaultCenter={center}
        onIdle={this.handleCenterChanged}
      > 
        { this.props.points &&
        <Polyline
          path={this.props.pathCoords}
          options={{
            strokeColor: '#FF0000',
            strokeOpacity: 1.0
          }}
        /> 
        }
        {
          this.props.points &&
          this.props.points.map((point, index) => {
            return (
              <Marker 
                key={point.pointId}
                position={point.latLng} 
                draggable={true}
                onDragStart={this.onStartChangePosition}
                onDragEnd={(event) => this.props.onChangePosition(point.pointId, event.latLng)}
                onClick={this.handleClickedMarker(point)}
              />
            )
          }) 
        }
        { 
          this.state.showInfoBox && <InfoWindow
            key={Math.random()}
            onCloseClick={this.handleCloseInfoBox}
            position={this.state.selectPoint.latLng}
          >
            <p>{this.state.selectPoint.pointName}</p>
          </InfoWindow>  
        } 
      </GoogleMap>
    )
  }
};

const MyMapComponent = compose(
  withProps({
    loadingElement: <div style={{ height: `100%` }} />,
    containerElement: <div style={{ 
                        height: `400px`,
                        border: `1px solid green`,
                        width: `66%`
                      }} />,
    mapElement: <div style={{ height: `100%` }} />,
    className: 'map-component'
  }),
  withGoogleMap
)(Map);

export default MyMapComponent;