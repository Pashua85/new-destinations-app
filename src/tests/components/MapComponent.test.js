import React from 'react';
import { shallow } from 'enzyme';
import MapComponent from '../../components/MapComponent';
import points from '../utils/points';
import pathCoords from '../utils/pathCoords';


test('should render MapComponent correctly', () => {
  const wrapper = shallow(
    <MapComponent points={points} pathCoords={pathCoords} />
  );
  expect(wrapper).toMatchSnapshot();
});