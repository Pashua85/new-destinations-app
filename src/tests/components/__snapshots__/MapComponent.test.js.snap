// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MapComponent correctly 1`] = `
<withGoogleMap(Map)
  className="map-component"
  containerElement={
    <div
      style={
        Object {
          "border": "1px solid green",
          "height": "400px",
          "width": "66%",
        }
      }
    />
  }
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  pathCoords={
    Array [
      Object {
        "lat": 56.80950300000001,
        "lng": 60.606506999999965,
      },
      Object {
        "lat": 56.8073729,
        "lng": 60.62031009999998,
      },
      Object {
        "lat": 56.854948,
        "lng": 60.661017000000015,
      },
    ]
  }
  points={
    Array [
      Object {
        "latLng": Object {
          "lat": 56.80950300000001,
          "lng": 60.606506999999965,
        },
        "pointId": "ChIJc8sXNttuwUMRWSTSqEBCxrY",
        "pointName": "Point one",
      },
      Object {
        "latLng": Object {
          "lat": 56.8073729,
          "lng": 60.62031009999998,
        },
        "pointId": "ChIJASguT89uwUMRMbx0H8V2uNk",
        "pointName": "Point two",
      },
      Object {
        "latLng": Object {
          "lat": 56.854948,
          "lng": 60.661017000000015,
        },
        "pointId": "ChIJ6TiVJjVswUMRRi8Evo28BM4",
        "pointName": "Point three",
      },
    ]
  }
/>
`;
